x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/evaluation/classification/MulticlassConfusionMatrix.java

```diff
@@ -0,0 +1,276 @@
+ * for all the possible class pairs {X, Y}.
+ */
+public class MulticlassConfusionMatrix implements ClassificationMetric {
+
+    public static final ParseField NAME = new ParseField("multiclass_confusion_matrix");
+
+    public static final ParseField SIZE = new ParseField("size");
+
+    private static final ConstructingObjectParser<MulticlassConfusionMatrix, Void> PARSER = createParser();
+
+    private static ConstructingObjectParser<MulticlassConfusionMatrix, Void> createParser() {
+        ConstructingObjectParser<MulticlassConfusionMatrix, Void>  parser =
+            new ConstructingObjectParser<>(NAME.getPreferredName(), true, args -> new MulticlassConfusionMatrix((Integer) args[0]));
+        parser.declareInt(optionalConstructorArg(), SIZE);
+        return parser;
+    }
+
+    public static MulticlassConfusionMatrix fromXContent(XContentParser parser) {
+        return PARSER.apply(parser, null);
+    }
+
+    private static final String STEP_1_AGGREGATE_BY_ACTUAL_CLASS = NAME.getPreferredName() + "_step_1_by_actual_class";
+    private static final String STEP_2_AGGREGATE_BY_ACTUAL_CLASS = NAME.getPreferredName() + "_step_2_by_actual_class";
+    private static final String STEP_2_AGGREGATE_BY_PREDICTED_CLASS = NAME.getPreferredName() + "_step_2_by_predicted_class";
+    private static final String STEP_2_CARDINALITY_OF_ACTUAL_CLASS = NAME.getPreferredName() + "_step_2_cardinality_of_actual_class";
+    private static final String OTHER_BUCKET_KEY = "_other_";
+    private static final int DEFAULT_SIZE = 10;
+    private static final int MAX_SIZE = 1000;
+
+    private final int size;
+    private List<String> topActualClassNames;
+    private Result result;
+
+    public MulticlassConfusionMatrix() {
+        this((Integer) null);
+    }
+
+    public MulticlassConfusionMatrix(@Nullable Integer size) {
+        if (size != null && (size <= 0 || size > MAX_SIZE)) {
+            throw ExceptionsHelper.badRequestException("[{}] must be an integer in [1, {}]", SIZE.getPreferredName(), MAX_SIZE);
+        }
+        this.size = size != null ? size : DEFAULT_SIZE;
+    }
+
+    public MulticlassConfusionMatrix(StreamInput in) throws IOException {
+        this.size = in.readVInt();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME.getPreferredName();
+    }
+
+    @Override
+    public String getName() {
+        return NAME.getPreferredName();
+    }
+
+    public int getSize() {
+        return size;
+    }
+
+    @Override
+    public final List<AggregationBuilder> aggs(String actualField, String predictedField) {
+        if (topActualClassNames == null) {
+            return List.of(
+                AggregationBuilders.terms(STEP_1_AGGREGATE_BY_ACTUAL_CLASS)
+                    .field(actualField)
+                    .order(List.of(BucketOrder.count(false), BucketOrder.key(true)))
+                    .size(size));
+        } else if (result == null) {
```

Comment:
We generally prefer that an `else if` clause is operating on the same data as the `if` clause to keep reasoning simple. In this case I think we don't really need the `else` at all as we are returning from each branch. So we could have:

```
if (topActualClassNames == null) {
    // This is step 1
    ...
    return {};
}
if (result == null) {
    // This is step 2
    ...
    return {};
}
return List.of();
```

Reply:
I wonder if it would also make it more readable to extract the building of the aggregations of each step into its own method.

Reply:
> In this case I think we don't really need the else at all as we are returning from each branch. 

Done.

> I wonder if it would also make it more readable to extract the building of the aggregations of each step into its own method.

I'm not sure this would help much. The method is not (yet) very long so I'd keep it as it is if you don't have a strong opinion about refactor.